src/critics/local_goal_critic.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 @artofnothingness Alexey Budyakov, Samsung Research
+#include "mppic/critics/local_goal_critic.hpp"
+
+#include <xtensor/xmath.hpp>
+#include <xtensor/xsort.hpp>
+
+namespace mppi::critics
+{
+
+void LocalGoalCritic::initialize()
+{
+  auto getParam = parameters_handler_->getParamGetter(name_);
+
+  getParam(consider_angle_, "consider_angle", true);
+  getParam(angle_offset_, "angle_offset", 6);
+  getParam(angle_cost_power_, "angle_cost_power", 1);
+  getParam(angle_cost_weight_, "angle_cost_weight", 3.0);
+
+  getParam(consider_distance_, "consider_distance", true);
+  getParam(distance_goal_count_, "distance_goal_count", 2);
+  getParam(distance_offset_, "distance_offset", 6);
+  getParam(distance_cost_power_, "distance_cost_power", 2);
+  getParam(distance_cost_weight_, "distance_cost_weight", 3.0);
+
+  getParam(stop_usage_path_reached_ratio_, "stop_usage_path_reached_ratio", 0.35);
+}
+
+void LocalGoalCritic::score(models::CriticFunctionData & data)
+{
+  if(!enabled_) {
+    return;
+  }
+
+  auto path_points = xt::view(data.path, xt::all(), xt::range(0, 2));
+
+  // get furthest point among trajectories nearest points to path
+  size_t furthest_reached_path_point_idx;
+  if (data.furthest_reached_path_point) {
+    furthest_reached_path_point_idx = *data.furthest_reached_path_point;
+  } else {
+    furthest_reached_path_point_idx = findPathFurthestPoint(data);
```

Comment:
If you make a single good implementation of `findPathFurthestPoint`, we could add it to the base class of the critics and then all critics can use it. Then we can basically have:

```
  if (!data.furthest_reached_path_point) {
    data.furthest_reached_path_point = findPathFurthestPoint(data);  // deal with pointer still
  } 
```

in any critic that might need it. It seems like if we decouple things like I suggest below with `PathAngle` that the path align / path angle / another local goal critic would all need it. With 3 total instances, that's worth abstracting to a method they can all share

Reply:
That's a nice idea

Reply:
done